### ISSUE TITLE
Respect LIBS environment variable

### ISSUE THREAD

**Issue by matoro** [mesonbuild/meson#10621]
[GNU autoconf uses the `LIBS` environment variable to specify additional libraries to link.](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/autoconf.html#index-LIBS-5).  This variable is distinct from `LDFLAGS`.  The latter is for flags that affect the behavior of the linker, while the former is for libraries to link.  This means that `LDFLAGS` is free to be placed before the object files in the command line, while `LIBS` must be placed after the object files.  The manual is very explicit about this in the description for `LDFLAGS`:
> Don’t use this variable to pass library names (`-l`) to the linker; use `LIBS` instead. 

There is a use-case for this.  On some architectures, like riscv, i386, and 32-bit sparc and mips, the [GCC __atomic intrinsics](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) are implemented in software via the libatomic library.  This means that packages using those intrinsics must link with with `-latomic`, but only on certain platforms.  On such platforms, the user will often need to set `LIBS="-latomic"` to successfully link.  This is easy to do with autotools-based projects, but meson does not know about this environment variable.

**Comment by eli-schwartz** [comment]
Wasn't there some GCC discussion about adding `--push-state --as-needed -latomic --pop-state` to the builtin spec on those platforms? This seems like a considerably more robust approach.

**Comment by matoro** [comment]
> Wasn't there some GCC discussion about adding `--push-state --as-needed -latomic --pop-state` to the builtin spec on those platforms? This seems like a considerably more robust approach.

Perhaps, and I can definitely look into this if you have a link.  However this isn't necessarily only the use case - just the particular one that prompted this issue.  A change to the gcc specs would take quite a long time before trickling down across the ecosystem.  A meson patch would be much faster.

**Comment by eli-schwartz** [comment]
Found some discussion:

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81358
but mostly https://github.com/riscv-collab/riscv-gcc/issues/12

I'm referring to option 4. It apparently was originally implemented, but later limited to whenever `-pthread` was passed, leading to the confusing situation where programs that use both threads and atomics worked, but programs that use only atomics did not work.

Also you're suggesting Meson implement a feature request to enable option 5. ;)

**Comment by dcbaker** [comment]
We handle a atomics in mesa and it’s a pain. Regardless of what we do with LIBS I think a built in dependency for atomics might be in order

**Comment by thesamesam** [comment]
See also https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8756.

**Comment by eebssk1** [comment]
So How do I add addition drop-in libraries to end of meson's linking command?
Is there anyway without modifying the build file?

**Comment by dcbaker** [comment]
@eebssk1 `meson setup $builddir -Dc_link_args=...` (also works with `meson configure`), use `<lang>_link_args`, so `cpp_link_args`, `objc_link_args`, etc.

**Comment by eebssk1** [comment]
> @eebssk1 `meson setup $builddir -Dc_link_args=...` (also works with `meson configure`), use `<lang>_link_args`, so `cpp_link_args`, `objc_link_args`, etc.

Hi.
I tried this.
However the flags are added before the actual (start-group objs stop-group) when final exe linking.

I checked the meson config log, the flags indeed appear at the end of the link command though.

**Comment by eebssk1** [comment]
> > @eebssk1 `meson setup $builddir -Dc_link_args=...` (also works with `meson configure`), use `<lang>_link_args`, so `cpp_link_args`, `objc_link_args`, etc.
> 
> Hi. I tried this. However the flags are added before the actual (start-group objs stop-group) when final exe linking.
> 
> I checked the meson config log, the flags indeed appear at the end of the link command though.

@dcbaker 
I think I found the problem here.
When building a executable, its dependencies reside in $LINK_ARGS instead of $in.
I think this is by design. However, passed <lang>_link_args come before the dependencies.
This behavior is inconsistent between what you told/the config log and the actual link.

**Comment by eebssk1** [comment]
> > > @eebssk1 `meson setup $builddir -Dc_link_args=...` (also works with `meson configure`), use `<lang>_link_args`, so `cpp_link_args`, `objc_link_args`, etc.
> > 
> > 
> > Hi. I tried this. However the flags are added before the actual (start-group objs stop-group) when final exe linking.
> > I checked the meson config log, the flags indeed appear at the end of the link command though.
> 
> @dcbaker I think I found the problem here after checking build.ninja. When building a executable, its dependencies reside in $LINK_ARGS instead of $in. I think this is by design. However, passed _link_args come before the dependencies. This behavior is inconsistent between what you told/the config log and the actual link.

For quick and dirty workaround that people may find useful for such situation. I added the flags manually at end of "rule c_LINKER ->  command" and the build works as intented.

**Comment by dcbaker** [comment]
@eebssk1 right, that's because if they're found in the meson.build file Meson knows what they are and handles them accordingly. the `c_link_args` are just a string list of flags that Meson attaches to the end of the link command, it doesn't know anything about them.

I've been meaning to look at this particular series again, since we do some of this dance in Mesa, but it's kinda a pain to know if we need `-latomic`

**Comment by eli-schwartz** [comment]
https://gcc.gnu.org/git/?p=gcc.git;a=log;h=e2f6ed54f75bbf8dd0292af90304890f06a9be17

The specific case of -latomic will be solved in GCC 15 (it has been reverted in the intervening hours since it caused a build regression but is expected to land again shortly).

This doesn't answer whether we should have a generic approach here. It also doesn't solve this for people running older compilers such as the most recently released one :) but there is light at the end of the tunnel...
